### PR TITLE
Enable Gemma4 VLM text batch admission

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Auto-enable the Python VLM text-only batch generator for safe Gemma4 `mlx_vlm` chat completions so imported Gemma4 text-only greedy requests can enter multimodal continuous-batch admission.
- Keep constrained requests out of the automatic path: non-greedy sampling, structured output, OpenAI tools, `tool_choice`, tool parser metadata, reasoning, thinking controls, and media still require explicit capability support.
- Document the OMLX/Melix benchmark finding and the Gemma4 auto-admission eligibility boundary in the serving benchmark plan.

## Plan or Spec
- `docs/plans/2026-05-12-omlx-melix-serving-benchmark.md`

## Commands Run
```text
git merge --ff-only origin/main
Result: passed; branch fast-forwarded to d8b1d8fb.

git diff --check
Result: passed.

swift test --package-path services/control-plane-swift --filter 'OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4TextOnlyVLMBatchGeneratorAdmission|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForNonGreedyRequests|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4VLMBatchGeneratorWithExplicitGreedyTopP|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForStructuredOutputs|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForOpenAITools|OpenAIHandlerTests/postChatCompletionsCopiesImportedVLMBatchGeneratorMetadataByRouteMetadata|OpenAIHandlerTests/postChatCompletionsLazyLoadsTextOnlyVLMRequestsThroughPythonVLMRoute|RequestCoordinatorTests/textOnlyPythonVLMBatchGeneratorRequestsCanEnterMultimodalContinuousBatchAdmission|RequestCoordinatorTests/pythonVLMRequestsWithMediaDoNotUseCooperativeTextBatching'
Result: passed; 9 tests in 2 suites.

make swift-test
Result: passed; Swift protocol, text-worker, control-plane, worker-client, and macOS menubar gates completed. Existing third-party deprecation, Swift Sendable, and CoreData NSXPCConnection warnings were emitted.

make py-test
Result: passed via pre-commit; 2879 passed, 14 skipped, 2 warnings in 149.86s.

make integration-test
Result: passed via pre-commit; 114 passed, 1 skipped in 657.65s.

pre-commit PR scoped performance report
Result: passed; Status ok, changed files 3, selected probes 0, direct/gated probes 0, regressions 0, context regressions 0, verification failures 0. Report: .runtime/pre-commit-performance/20260520-083323-d8b1d8fb/report/report.md

UV_PROJECT_ENVIRONMENT=/Users/chenyu/Documents/github/melix/.runtime/worktrees/omlx-melix-benchmark-20260520/.venv uv run --frozen python scripts/omlx_melix_compare_benchmark.py --models unsloth/gemma-4-31b-it-UD-MLX-4bit --omlx-model gemma-4-31b-it-UD-MLX-4bit --melix-base-url http://127.0.0.1:12452/v1 --omlx-base-url http://127.0.0.1:18000/v1 --concurrency 4 --prompt-tokens 128,512 --max-tokens 64 --warmups 2 --samples 4
Result: passed; exported /Users/chenyu/Downloads/latest-omlx-melix-gemma31b-c4-autobatch-20260520-164147 with 32 observations and 0 errors.

Review refresh after code review:

git merge --no-edit origin/main
Result: passed; merged origin/main at 28eccf21 into the PR branch.

git diff --check
Result: passed after review updates.

swift test --package-path services/control-plane-swift --filter 'OpenAIHandlerTests/postChatCompletionsLazyLoadsTextOnlyVLMRequestsThroughPythonVLMRoute|OpenAIHandlerTests/postChatCompletionsCopiesImportedVLMBatchGeneratorMetadataByRouteMetadata|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4TextOnlyVLMBatchGeneratorAdmission|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForNonGreedyRequests|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4VLMBatchGeneratorWithExplicitGreedyTopP|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForStructuredOutputs|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForOpenAITools'
Result: passed; 7 OpenAIHandler VLM tests in 1 suite.

Second main refresh:

git merge --no-edit origin/main
Result: passed; merged origin/main at 6d9faa88 into the PR branch with no conflicts.

git diff --check
Result: passed after the second main refresh.

Third main refresh:

git merge --no-edit origin/main
Result: passed; merged origin/main at 20523388 into the PR branch with no conflicts.

git diff --check
Result: passed after the third main refresh.

Review refresh for explicit non-greedy VLM batch-generator metadata:

swift test --package-path services/control-plane-swift --filter 'OpenAIHandlerTests/postChatCompletionsCopiesImportedVLMBatchGeneratorMetadataByRouteMetadata|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4TextOnlyVLMBatchGeneratorAdmission|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForNonGreedyRequests|OpenAIHandlerTests/postChatCompletionsDoesNotNormalizeExplicitVLMBatchGeneratorNonGreedySampling|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4VLMBatchGeneratorWithExplicitGreedyTopP|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForStructuredOutputs|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForOpenAITools'
Result: passed; 7 OpenAIHandler VLM tests in 1 suite.

Fourth main refresh:

git merge --no-edit origin/main
Result: passed; merged origin/main at b7886e1f into the PR branch with no conflicts.

git diff --check
Result: passed after the fourth main refresh.

swift test --package-path services/control-plane-swift --filter 'OpenAIHandlerTests/postChatCompletionsCopiesImportedVLMBatchGeneratorMetadataByRouteMetadata|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4TextOnlyVLMBatchGeneratorAdmission|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForNonGreedyRequests|OpenAIHandlerTests/postChatCompletionsDoesNotNormalizeExplicitVLMBatchGeneratorNonGreedySampling|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4VLMBatchGeneratorWithExplicitGreedyTopP|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForStructuredOutputs|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForOpenAITools'
Result: passed; 7 OpenAIHandler VLM tests in 1 suite.

CI failure follow-up for explicit non-greedy sampling assertion:

git diff --check
Result: passed after changing the explicit non-greedy test to use imported top_p=0.8.

swift test --package-path services/control-plane-swift --filter 'OpenAIHandlerTests/postChatCompletionsCopiesImportedVLMBatchGeneratorMetadataByRouteMetadata|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4TextOnlyVLMBatchGeneratorAdmission|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForNonGreedyRequests|OpenAIHandlerTests/postChatCompletionsDoesNotNormalizeExplicitVLMBatchGeneratorNonGreedySampling|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4VLMBatchGeneratorWithExplicitGreedyTopP|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForStructuredOutputs|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForOpenAITools'
Result: passed; 7 OpenAIHandler VLM tests in 1 suite.

Fifth main refresh after PR #1319:

git fetch origin main codex/omlx-melix-benchmark-20260520
Result: passed; origin/main advanced to eee6b713 from PR #1319.

git merge --no-edit origin/main
Result: passed; merged origin/main at eee6b713 into the PR branch with no conflicts.

git diff --check
Result: passed after the fifth main refresh.

swift test --package-path services/control-plane-swift --filter 'OpenAIHandlerTests/postChatCompletionsCopiesImportedVLMBatchGeneratorMetadataByRouteMetadata|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4TextOnlyVLMBatchGeneratorAdmission|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForNonGreedyRequests|OpenAIHandlerTests/postChatCompletionsDoesNotNormalizeExplicitVLMBatchGeneratorNonGreedySampling|OpenAIHandlerTests/postChatCompletionsAutoEnablesGemma4VLMBatchGeneratorWithExplicitGreedyTopP|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForStructuredOutputs|OpenAIHandlerTests/postChatCompletionsKeepsGemma4VLMBatchGeneratorDisabledForOpenAITools'
Result: passed; 7 OpenAIHandler VLM tests in 1 suite.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python --extra mlx pytest services/mlx-worker-python/tests/test_mlx_vlm_runtime.py -q
Result: passed; 69 passed in 0.44s.
```

## Coverage and Metrics
- Changed scope: Swift HTTP gateway/control-plane VLM admission.
- Coverage: focused admission coverage is provided by the 9-test Swift filter above plus the full `make swift-test`, `make py-test`, and `make integration-test` gates. The current Swift workflow does not emit a line/branch coverage percentage for this scope, so a numeric 95% coverage report is not currently measurable from the existing command surface.
- Observability mode: evidence. This PR does not add new production probes; benchmark evidence reused existing scheduler and VLM metric snapshots.
- Probe overhead: N/A for this PR because no new runtime metrics or evidence probes were added.
- PR scoped performance: `Status: ok`; no registered performance probes were selected for this change set and no regressions or verification failures were reported.
- Metrics, c=4 smoke bundle `/Users/chenyu/Downloads/latest-omlx-melix-gemma31b-c4-autobatch-20260520-164147`:
  - Melix 128-token prompt, c=4, max64: TTFT 4014.83 ms, total 6999.94 ms, decode 21.44 tok/s, aggregate 27.57 tok/s.
  - OMLX 128-token prompt, c=4, max64: TTFT 3196.10 ms, total 8194.60 ms, decode 12.79 tok/s, aggregate 30.95 tok/s.
  - Melix 512-token prompt, c=4, max64: TTFT 6511.03 ms, total 9558.72 ms, decode 21.01 tok/s, aggregate 20.22 tok/s.
  - OMLX 512-token prompt, c=4, max64: TTFT 6403.56 ms, total 11484.91 ms, decode 12.61 tok/s, aggregate 22.15 tok/s.
  - Melix/OMLX aggregate ratio: 0.891 for 128-token prompts, 0.913 for 512-token prompts.
  - Melix scheduler evidence: `scheduler.multimodal_continuous_batch_enabled=1`, `scheduler.multimodal_continuous_batch_effective_capacity=2`, `scheduler.continuous_batch_eligible_rate=100`, `scheduler.continuous_batch_merge_rate=47.0588`, `vision.text_batch_generator.submitted_request_count=17`, `completed_request_count=17`, `peak_active_batch_size=2`, `failed_request_count=0`.

## Known Gaps
- `make bootstrap` and `make proto` were not rerun for this PR. No bootstrap, protocol schema, or generated artifact changes were made.
- No protocol schema, dependency, or generated artifact changes were made.
- No larger model matrix was rerun; this PR records the targeted Gemma4 31B c=4 evidence and keeps the broader benchmark sweep as plan work.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts. N/A: no protocol changes.
- [x] Dependency changes include updated lockfiles. N/A: no dependency changes.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and any deferred debug or evidence work are recorded, or `N/A` is stated explicitly with the reason.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift [hunk 1]
```diff
@@ -2057,6 +2062,39 @@ public struct OpenAIHandler: Sendable {
...
+              normalizedRequest.toolParser == nil,
+              normalizedRequest.tools.isEmpty,
+              normalizedRequest.toolChoice == nil,
+              normalizedRequest.reasoningEffort == nil,
+              normalizedRequest.enableThinking != true,
+              normalizedRequest.thinking?.isEnabled != true
+        else {
+            return false
+        }
+        guard let requestedTemperature = normalizedRequest.temperature,
```
Comment (Zigfreidish) [inline services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift]: The dual epsilon thresholds (`1e-9` for `requestedTemperature` vs `1e-6` for `workerRequest.sampling.temperature`) are intentional but silent. `workerRequest.sampling.temperature` is a Float32 protobuf scalar, so the Double-cast can carry rounding error at ~7 decimal places — `1e-6` is the right bound for that. A brief comment here would save the next reader from having to re-derive this:

```swift
// workerRequest.sampling.temperature is Float32; 1e-6 accommodates the Float→Double conversion loss
abs(Double(workerRequest.sampling.temperature)) < 1e-6
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift [hunk 2]
```diff
@@ -2057,6 +2062,39 @@ public struct OpenAIHandler: Sendable {
...
+              abs(requestedTemperature) < 1e-9,
+              abs(Double(workerRequest.sampling.temperature)) < 1e-6
+        else {
+            return false
```
Comment (Zigfreidish) [inline services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift:2093]: Worth considering whether `normalizedRequest.topK` needs a guard here. A request with `temperature=0, top_k=40` would currently pass all checks and get auto-admitted, but top-k sampling is non-greedy. If `mlx_vlm` respects `top_k` in the batch path, that would be a safety gap. If `top_k` is ignored or clamped to 0 in this code path, a comment saying so would close the question definitively.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
Solid PR. The admission gate logic is correct and well-scoped: the eligibility rule precisely matches the documented boundary (explicit greedy, no constrained outputs, no tools, no reasoning), and the 9-test suite covers the key positive and negative cases cleanly.

Two nits inline — both minor and non-blocking:
1. The `1e-9` vs `1e-6` epsilon asymmetry is defensible (Float32 conversion headroom) but would benefit from a one-line comment.
2. No `top_k` guard — worth a quick note confirming whether mlx_vlm ignores it in this path.

Everything else looks good: the `|| shouldAutoEnableVLMTextOnlyBatchGenerator(...)` short-circuit composes correctly with the existing explicit-flag path, the doc update captures the eligibility boundary accurately, and the benchmark metrics are well-evidenced.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/control-plane-swift/Tests/HTTPGatewayTests/OpenAIHandlerTests.swift [hunk 3]
```diff
@@ -1374,6 +1374,353 @@ struct OpenAIHandlerTests {
         #expect(await textClient.lastGenerateRequest == nil)
     }
 
+    @Test("POST /v1/chat/completions auto-enables Gemma4 text-only VLM batch-generator admission")
+    func postChatCompletionsAutoEnablesGemma4TextOnlyVLMBatchGeneratorAdmission() async throws {
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/control-plane-swift/Tests/HTTPGatewayTests/OpenAIHandlerTests.swift:1378]: Done in acd2a050: extracted makeGemma4VLMOpenAIHandler and reused it across the Gemma4 VLM admission tests so each case now focuses on request shape and assertions.

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift]: Done in acd2a050: simplified the auto-admission temperature check to require an explicit request temperature plus the final worker sampling temperature within Float32-sized tolerance.

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift (hunk 1 above)
Comment (Zigfreidish) [reply services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift]: Done in acd2a050: removed the dual epsilon check and left the tolerance comment on the final worker Float32 sampling value.

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift (hunk 2 above)
Comment (Zigfreidish) [reply services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift:2093]: Done in acd2a050: added a final worker sampling.topK == 0 guard before auto-admission and documented that the gate applies after request shaping, so future top-k shaping cannot enter this path.

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift
Comment (Zigfreidish) [inline services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift:1]: **`shouldNormalizeVLMTextOnlyBatchGeneratorSampling` — relaxed temperature guard**

The old guard was `abs(requestedTemperature) < 1e-9`; the new one is just `temperature != nil`. For the new auto-enable code path this is correct — `shouldAutoEnableVLMTextOnlyBatchGenerator` already enforces greedy sampling at the worker level before this is reached. However, the explicit opt-in path (`melix.vlm.text_only_batch_generator = true` in model metadata) no longer checks that the incoming temperature is zero before normalising. The function still falls back to checking `abs(Double(workerRequest.sampling.temperature)) < 1e-6` on the last line, so the end result is safe today. Worth adding a brief comment or test that covers the explicit opt-in + non-zero temperature case so the intent stays clear if the sampling shaping order changes later.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
Overall the implementation is clean and well-tested. The admission gate in `shouldAutoEnableVLMTextOnlyBatchGenerator` is thorough — it guards on backend, vision family, sampling mode, structured output, tool use, and reasoning controls before auto-enabling. The 9 Swift tests cover the positive path and all the key exclusion cases. The benchmark evidence and scheduler metrics in the PR description are a solid reference for the eligibility rule.

One nit inline about the relaxed temperature guard in `shouldNormalizeVLMTextOnlyBatchGeneratorSampling` — safe today but worth a clarifying comment or test to preserve intent. **LGTM — would approve if GitHub allowed self-review.**

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift
Comment (Zigfreidish) [reply services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift:1]: Added a focused regression test for this explicit metadata path: `postChatCompletionsDoesNotNormalizeExplicitVLMBatchGeneratorNonGreedySampling` now verifies that `melix.vlm.text_only_batch_generator=true` is still copied to the worker request while a non-zero requested temperature is preserved and does not trigger top-p/top-k normalization. The adjacent VLM admission filter passes after the change.

**Review by Zigfreidish — COMMENTED** [review]
**Review — ready to merge**

Clean implementation with correct eligibility guards.

---

**`OpenAIHandler.swift` — `shouldAutoEnableVLMTextOnlyBatchGenerator`**

The eligibility fence is thorough: Gemma4 `mlx_vlm` only, no structured output, no tool parser, no tools, no tool choice, no reasoning/thinking, and a two-layer temperature check. Worth calling out the design intent explicitly:

- `normalizedRequest.temperature != nil` requires the caller to have *explicitly* sent `temperature: 0`, not rely on a server-side default. A request that omits `temperature` entirely and receives a backend default of `0` does not auto-enter the batch path. This is the right conservative boundary.
- `abs(Double(workerRequest.sampling.temperature)) < 1e-6` on the *post-shaped* worker request prevents any future sampling normalizer from silently widening the greedy window.

Both layers should be preserved if this path is ever extended.

---

**`shouldNormalizeVLMTextOnlyBatchGeneratorSampling` simplification**

The old near-zero check (`abs(requestedTemperature) < 1e-9`) on the normalization path was stranded — `shouldAutoEnableVLMTextOnlyBatchGenerator` now enforces `< 1e-6` on the worker value. Removing it from the normalization gate and leaving only `temperature != nil` is correct. Greedy enforcement lives in the right place.

---

**Test helper `makeGemma4VLMOpenAIHandler`**

Seeding `melix.vlm.text_only_batch_generator = "false"` correctly simulates the imported Gemma4 state that triggered the production bottleneck. The `configureModel` closure gives individual tests the flexibility to override metadata without duplicating harness setup.

---

**Benchmark evidence**

Melix decode throughput (21.44 tok/s) vs OMLX (12.79 tok/s) at c=4, 128-token prompt is a meaningful signal. `continuous_batch_eligible_rate=100` and `continuous_batch_merge_rate=47` confirm requests are entering the batch path as intended, with zero failed requests.

---

Implementation is safe, the 7-test suite covers all stated eligibility boundaries, and the benchmark evidence validates the intended behavior. ✅

---
_Generated by [Claude Code](https://claude.ai/code)_
